FloripaPublicTransportation/MapViewController.swift

```diff
@@ -0,0 +1,140 @@
+    /// during the lifetime of `MapViewController`
+    private func showAndUpdateUserLocation() {
+        self.locationManager.startUpdatingLocation()
+        self.mapView.showsUserLocation = true
+    }
+    
+    /// NSTimer callback to show the pin callout automatically once it is created
+    internal func showPinCallout() {
+        // can't be private otherwise the selector does not recognize it
+        mapView.selectAnnotation(mapView.annotations.filter({ (a: MKAnnotation) -> Bool in
+            return !a.isKindOfClass(MKUserLocation) }).first!, animated: true)
+    }
+    
+    // MARK: - Delegates
+    
+    // MARK: CLLocationManagerDelegate
+    
+    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
+        if status == CLAuthorizationStatus.AuthorizedAlways || status == CLAuthorizationStatus.AuthorizedWhenInUse {
+            self.showAndUpdateUserLocation()
+        }
+    }
+    
+    // MARK: MKMapViewDelegate
+    
+    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
+        guard !annotation.isKindOfClass(MKUserLocation) else {
+            return nil
+        }
+        
+        let annotationView = MKPinAnnotationView()
+        annotationView.annotation = annotation
+        annotationView.canShowCallout = true
+        annotationView.pinTintColor = UIColor.redColor()
+        
+        let okButton = UIButton(type: .System)
+        okButton.setTitle("Ok", forState: .Normal)
+        okButton.frame = CGRectMake(0, 0, annotationView.frame.width, annotationView.frame.height)
+        annotationView.rightCalloutAccessoryView = okButton
+        
+        return annotationView
+    }
+    
+    func mapView(mapView: MKMapView, didAddAnnotationViews views: [MKAnnotationView]) {
+        /// zooms the map to the user location on first load
+        if mapView.annotations.count == 1 && mapView.annotations.first!.isKindOfClass(MKUserLocation) {
+            mapView.showAnnotations(mapView.annotations, animated: true)
+            return
+        }
+        else if mapView.annotations.count >= 1  {
+            // for some reason, the callout is not shown if the line below is called here:
+            //mapView.selectAnnotation(mapView.annotations.last!, animated: false)
+            
+            // will have to adopt an ugly way and hardcode a time interval that always works:
+            _ = NSTimer.scheduledTimerWithTimeInterval(0.8, target: self, selector: "showPinCallout", userInfo: nil, repeats: false)
+        }
+    }
+    
+    func mapView(mapView: MKMapView, annotationView view: MKAnnotationView, calloutAccessoryControlTapped control: UIControl) {
+        guard self.navigationController != nil else {
+            return
+        }
+        
+        if let routesListVC = self.navigationController!.viewControllers[0] as? RoutesListViewController {
```

Comment:
Oh, I see... Yes, for maintainability your suggestion would be better. The required application wasn't big, but the time I had to spend on it wasn't enough to take of all micro details - I've really preferred to invest the time I had on adding bigger impact extras like unit tests, spinners, connectivity checks, sentence capitalization, among others. I'm pretty sure this is not a perfect maintainable app heheh but I really do appreciate your comments and effort on going very deep through it :)